crates/matrix-sdk-indexeddb/src/state_store/mod.rs

```diff
@@ -623,10 +624,23 @@ impl_state_store!({
 
         if !changes.room_infos.is_empty() {
             let room_infos = tx.object_store(keys::ROOM_INFOS)?;
+
             for (room_id, room_info) in &changes.room_infos {
+                // Load raw room info from db, add our updates and then save the raw version again.
+
+                let mut raw_room_info = room_infos
+                    .get(&self.encode_key(keys::ROOM_INFOS, room_id))?
+                    .await?
+                    .and_then(|v| self.deserialize_event::<RawRoomInfo>(&v).ok())
+                    .unwrap_or_else(|| room_info.clone().into_raw_lossy());
+
+                if let Some(room_state_changes) = changes.state.get(room_id) {
+                    raw_room_info.base_info.extend_with_changes(&room_state_changes);
+                }
```

Comment:
It seems it would be much simpler to throw away the previous `RawRoomInfo` and convert the `RoomInfo` struct into a `RawRoomInfo` struct, instead of applying a diff of changes to the previously-serialized `RawRoomInfo`? That way maybe we wouldn't even need `extend_with_changes` at all.